pennylane/gradients/param_shift_hessian.py

```diff
@@ -0,0 +1,195 @@
+        shifts (list[list[float or int]]): nested list of shift values, each
+            list containing a value for each index
+
+    Returns:
+        list[QuantumTape]: List of quantum tapes. Each tape has multiple parameters
+            (indicated by ``idx``) shifted by the values of ``shifts``. The length
+            of the returned list of tapes will match the length of ``shifts``.
+    """
+    params = list(tape.get_parameters())
+    tapes = []
+
+    for shift in shifts:
+        new_params = params.copy()
+        shifted_tape = tape.copy(copy_operations=True)
+
+        for i, s in enumerate(shift):
+            new_params[idx[i]] = new_params[idx[i]] + qml.math.convert_like(s, new_params[idx[i]])
+
+        shifted_tape.set_parameters(new_params)
+        tapes.append(shifted_tape)
+
+    return tapes
+
+
+class hessian_transform(qml.batch_transform):
+    def __init__(
+        self, transform_fn, expand_fn=expand_invalid_trainable, differentiable=True, hybrid=True
+    ):
+        self.hybrid = hybrid
+        super().__init__(transform_fn, expand_fn=expand_fn, differentiable=differentiable)
+
+    def default_qnode_wrapper(self, qnode, targs, tkwargs):
+        # Here, we overwrite the QNode execution wrapper in order
+        # to take into account that classical processing may be present
+        # inside the QNode.
+        hybrid = tkwargs.pop("hybrid", self.hybrid)
+        _wrapper = super().default_qnode_wrapper(qnode, targs, tkwargs)
+        cjac_fn = qml.transforms.classical_jacobian(qnode, expand_fn=expand_invalid_trainable)
+
+        def jacobian_wrapper(*args, **kwargs):
+            qjac = _wrapper(*args, **kwargs)
+
+            if any(m.return_type is qml.operation.Probability for m in qnode.qtape.measurements):
+                qjac = qml.math.squeeze(qjac)
+
+            if not hybrid:
+                return qjac
+
+            kwargs.pop("shots", False)
+            cjac = cjac_fn(*args, **kwargs)
+
+            if isinstance(cjac, tuple):
+                # Classical processing of multiple arguments is present. Return qjac @ cjac.
+                jacs = [
+                    qml.math.squeeze(qml.math.tensordot(c, qjac, [[0], [-1]]))
+                    for c in cjac
+                    if c is not None
+                ]
+                return jacs
+
+            is_square = cjac.shape == (1,) or (cjac.ndim == 2 and cjac.shape[0] == cjac.shape[1])
+
+            if is_square and qml.math.allclose(cjac, qml.numpy.eye(cjac.shape[0])):
+                # Classical Jacobian is the identity. No classical processing
+                # is present inside the QNode.
+                return qjac
+
+            # Classical processing of a single argument is present. Return cjac.T @ qjac @ cjac.
+            jac = qml.math.tensordot(qjac, cjac, [[-1], [0]])
+            jac = qml.math.tensordot(jac, cjac, [[-2], [0]])
+            return qml.math.swapaxes(jac, -1, -2)
```

Comment:
From playing around with this locally, I think this might still need to be tweaked to support arbitrary output sizes!

The example I was using was:

```python
dev = qml.device("default.qubit", wires=2)

@qml.beta.qnode(dev, diff_method="parameter-shift", max_diff=2)
def circuit(weights):
    qml.RX(weights[0, 0], wires=0)
    qml.RY(weights[0, 1], wires=0)
    qml.CNOT(wires=[0, 1])
    qml.RX(weights[0, 2], wires=0)
    qml.RY(weights[0, 0], wires=0)
    return qml.probs(wires=0), qml.probs(wires=1)

weights = np.ones([1, 3], requires_grad=True)
```

In this example, 

- The QNode argument shape is `(1, 3)`
- The number of gate arguments is `4`
- the output of shape of the QNode is `(2, 2)`

As a result, we have:

- Classical Jacobian with shape `(4, 1, 3)` (e.g., `(gate_args, qnode_args)`)
- QUantum circuit Hessian with shape `(2, 4, 4, 2)` (as currently returned by your function).

We need to somehow compute `J @ H @ J.T` such that the output matrix has shape `(2, 2, 1, 3, 1, 3)` (that is, `(output_shape, qnode_args, qnode_args)`).

```pycon
>>>  qml.jacobian(qml.jacobian(circuit))(weights).shape
(2, 2, 1, 3, 1, 3)
```

This might be easier if you could 'invert' the shape of your returned Hessian; e.g., if your circuit Hessian was returned with shape `(4, 2, 2, 4)` (so that the `gate_args` dimensions are always axes 0 and -1), then we could do:

```python
>>> hess = qml.math.tensordot(cjac, qhess, [[0], [0]])
>>> hess.shape
(1, 3, 2, 2, 4)
>>> hess = qml.math.tensordot(cjac, hess, [[0], [-1]])
>>> hess.shape
(1, 3, 1, 3, 2, 2)
```

... oh. We need to move the `(2, 2)` to the beginning 🤦 

This is a lot harder to puzzle out than I thought!

Reply:
Good catch! Those lines indeed only consider a classical Jacobian with 2 dimensions. Let me think about this a bit :)

Reply:
9916fea should resolve this issue. Are we guaranteed that the gate arguments are always 1 dimensional?

Reply:
I believe so, I cannot think of any examples where that is not the case!